### PR TITLE
i#5023: AArch64 decode: generate condition flags register r/w marking

### PR DESCRIPTION
**Pull request by AssadHashmi**
This patch adds a new field to codec.txt to make developers
explicitly specify whether an instruction touches the condition
code register.

For v8.0:
```
Read only      Write only      Read and write
adc            adds            adcs
bcond          ands            ccmn
csel           bics            ccmp
csinc          fccmp           sbcs
csinv          fccmpe
csneg          fcmp
fcsel          fcmpe
sbc            subs
mrs            msr
```

Issues: #5023, #5025

### CODE REVIEW

File: core/ir/aarch64/codec.txt [hunk 1]
```diff
@@ -212,1135 +212,1153 @@ x----------xxxxx----------------  wx16       # W/X register (or WZR/XZR)
...
+# instruction, i.e. whether the instruction can read from the register, write
+# to the register or both.
+
+# There are 6 states for nzcv_flag. The field should appear between the
+# instruction pattern and the opcode as shown above. The flag states are:
+# n No read or write
+# r Read only
+# w Write only
+# rw or wr Read and write
+# er Explicit/direct read (currently only MRS instruction)
```
Comment (derekbruening) [inline core/ir/aarch64/codec.txt:229]: Maybe mention that these require companion code in codec.c to set the instr_t.eflags field

File: core/ir/aarch64/codec.c [hunk 2]
```diff
@@ -3806,25 +3806,22 @@ decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
         instr->dsts[3] = opnd_create_reg(DR_REG_X0 + (enc >> 16 & 31));
```
Comment (derekbruening) [inline core/ir/aarch64/codec.c:3806]: style nit: s/i5023/i#5023/ in title

File: core/ir/aarch64/codec.txt [hunk 3]
```diff
@@ -1350,232 +1368,232 @@ x001111001000010xxxxxxxxxxxxxxxx     scvtf     d0 : wx5 scale
...
+0x0011101x100000111010xxxxxxxxxx  n     fcmlt     dq0 : dq5 sd_sz
+00011110111xxxxx001000xxxxx10000  w     fcmpe         : h5 h16
+0001111011100000001000xxxxx11000  w     fcmpe         : h5
+00011110001xxxxx001000xxxxx10000  w     fcmpe         : s5 s16
+0001111000100000001000xxxxx11000  w     fcmpe         : s5
+00011110011xxxxx001000xxxxx10000  w     fcmpe         : d5 d16
+0001111001100000001000xxxxx11000  w     fcmpe         : d5
+0111111001100001011010xxxxxxxxxx  n     fcvtxn     s0 : d5
+0010111001100001011010xxxxxxxxxx  n     fcvtxn     d0 : q5
+0110111001100001011010xxxxxxxxxx  n     fcvtxn2    q0 : q5
```
Comment (derekbruening) [inline core/ir/aarch64/codec.txt:1599]: The only simple test to add would be encoding some of these opcodes and querying the eflags field vs hardcoded expectations -- still, maybe worth adding a few such sanity checks to api.ir?

**Review by derekbruening — APPROVED** [review]
Thank you for this cleanup: I think this will avoid future problems when the codec is tweaked (e.g., for #4393) and when new opcodes are added.

File: core/ir/aarch64/codec.py [hunk 4]
```diff
@@ -103,10 +103,29 @@ def reorder_key(t):
...
-                c.append('%s    return decode_opnds%s(enc, dc, pc, instr, OP_%s);' %
-                         (indent, t, m))
+                if opc_props[m] == 'n':
+                    c.append('%sif ((enc & 0x%08x) == 0x%08x)' %
+                             (indent, ((1 << N) - 1) & ~v, f))
+                    c.append('%s    return decode_opnds%s(enc, dc, pc, instr, OP_%s);' %
+                             (indent, t, m))
+                else:
+                    c.append('%sif ((enc & 0x%08x) == 0x%08x) {' %
+                             (indent, ((1 << N) - 1) & ~v, f))
```
[the remark this answers is not in the document]
Comment (AssadHashmi) [reply core/ir/aarch64/codec.py]: Done.
There's a balance between clean Python generator code and clean generated source, which I think this update has done.

File: core/ir/aarch64/codec.txt (hunk 1 above)
Comment (AssadHashmi) [reply core/ir/aarch64/codec.txt:229]: Done.

File: core/ir/aarch64/codec.txt (hunk 3 above)
Comment (AssadHashmi) [reply core/ir/aarch64/codec.txt:1600]: Noted. Will do this as part of https://github.com/DynamoRIO/dynamorio/issues/5025.

**Pull request merged**